[Java client] `TypedApiEntityConsumer` fails to handle incoming data when chunk size in bytes is >2048

**Describe the bug**

<!-- A clear and concise description of what the bug is.
For existing Camunda Customers,
please also use our [Support Channels](https://camunda.com/services/support/) to
get priority assistance -->

When using the Camunda 8 REST API via the Java client, `TypedApiEntityConsumer` fails to handle the response body if the size of the received data chunk is more than 2048 bytes.

I tracked the issue down to this method: [`consumeData`](https://github.com/camunda/camunda/blob/03ed7054ca5046d41a1be670afd96a4551049c82/clients/java/src/main/java/io/camunda/zeebe/client/impl/http/TypedApiEntityConsumer.java#L175). Initially, the buffer size is 1024, and in that method we check if the incoming data fits in the buffer, and if not, we allocate an additional 1024 bytes. The problem is, if the payload chunk is e.g. 4 KB in size, even after the increase, the buffer will not be large enough to accept it.

I encountered this when calling the [Get process definition XML endpoint](https://docs.camunda.io/docs/next/apis-tools/camunda-api-rest/specifications/get-process-definition-xml/) - the XML string in response is easily longer than 2048 bytes.

**To Reproduce**

<!--
Steps to reproduce the behavior

If possible add a minimal reproducer code sample
- when using the Java client: https://github.com/zeebe-io/zeebe-test-template-java

-->

1. Ensure some process definitions are deployed
2. Create a Zeebe client with the default config and execute the process definition XML query for any existing process definition
3. Observe the error:
```
io.camunda.zeebe.client.api.command.ClientException: io.camunda.zeebe.client.api.command.ClientException: java.lang.IndexOutOfBoundsException: Range [0, 0 + 4279) out of bounds for length 2048
```

**Expected behavior**

<!-- A clear and concise description of what you expected to happen. -->

The body buffer should be increased not by 1024 bytes, but proportionally to the data chunk size - potentially with a limit equal to `maxMessageSize` from the Zeebe client config.

**Log/Stacktrace**

<!-- If possible add the full stacktrace or Zeebe log which contains the issue. -->

<details><summary>Full Stacktrace</summary>
 <p>

```
2024-11-15T16:02:09.303+01:00 ERROR 26112 --- [   scheduling-1] i.c.c.r.i.state.ProcessStateStoreImpl    : Failed to register process Process_0wxlkm3

io.camunda.zeebe.client.api.command.ClientException: io.camunda.zeebe.client.api.command.ClientException: java.lang.IndexOutOfBoundsException: Range [0, 0 + 4279) out of bounds for length 2048
	at io.camunda.zeebe.client.impl.http.HttpZeebeFuture.unwrapExecutionException(HttpZeebeFuture.java:86)
	at io.camunda.zeebe.client.impl.http.HttpZeebeFuture.join(HttpZeebeFuture.java:42)
	at io.camunda.connector.runtime.inbound.operate.OperateClientImpl.getProcessModel(OperateClientImpl.java:89)
	at io.camunda.connector.runtime.inbound.state.ProcessDefinitionInspector.findInboundConnectors(ProcessDefinitionInspector.java:89)
	at io.camunda.connector.runtime.inbound.state.ProcessStateStoreImpl.getConnectors(ProcessStateStoreImpl.java:153)
	at io.camunda.connector.runtime.inbound.state.ProcessStateStoreImpl.lambda$newlyDeployed$4(ProcessStateStoreImpl.java:89)
	at java.base/java.util.HashMap.compute(HashMap.java:1324)
	at io.camunda.connector.runtime.inbound.state.ProcessStateStoreImpl.newlyDeployed(ProcessStateStoreImpl.java:86)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at io.camunda.connector.runtime.inbound.state.ProcessStateStoreImpl.update(ProcessStateStoreImpl.java:77)
	at io.camunda.connector.runtime.inbound.state.TenantAwareProcessStateStoreImpl.lambda$update$2(TenantAwareProcessStateStoreImpl.java:56)
	at java.base/java.util.HashMap.forEach(HashMap.java:1429)
	at io.camunda.connector.runtime.inbound.state.TenantAwareProcessStateStoreImpl.update(TenantAwareProcessStateStoreImpl.java:45)
	at io.camunda.connector.runtime.inbound.importer.ProcessDefinitionImporter.handleImportedDefinitions(ProcessDefinitionImporter.java:83)
	at io.camunda.connector.runtime.inbound.importer.ProcessDefinitionImporter.scheduleImport(ProcessDefinitionImporter.java:56)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.springframework.scheduling.support.ScheduledMethodRunnable.runInternal(ScheduledMethodRunnable.java:130)
	at org.springframework.scheduling.support.ScheduledMethodRunnable.lambda$run$2(ScheduledMethodRunnable.java:124)
	at io.micrometer.observation.Observation.observe(Observation.java:499)
	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:124)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:358)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: io.camunda.zeebe.client.api.command.ClientException: java.lang.IndexOutOfBoundsException: Range [0, 0 + 4279) out of bounds for length 2048
	at io.camunda.zeebe.client.impl.http.ApiCallback.failed(ApiCallback.java:69)
	at org.apache.hc.core5.concurrent.BasicFuture.failed(BasicFuture.java:166)
	at org.apache.hc.core5.concurrent.ComplexFuture.failed(ComplexFuture.java:79)
	at org.apache.hc.client5.http.impl.async.InternalAbstractHttpAsyncClient$2.failed(InternalAbstractHttpAsyncClient.java:356)
	at org.apache.hc.client5.http.impl.async.AsyncRedirectExec$1.failed(AsyncRedirectExec.java:246)
	at org.apache.hc.client5.http.impl.async.AsyncHttpRequestRetryExec$1.failed(AsyncHttpRequestRetryExec.java:195)
	at org.apache.hc.client5.http.impl.async.AsyncProtocolExec$1.failed(AsyncProtocolExec.java:295)
	at org.apache.hc.client5.http.impl.async.HttpAsyncMainClientExec$1.failed(HttpAsyncMainClientExec.java:136)
	at org.apache.hc.core5.http.impl.nio.ClientHttp1StreamHandler.failed(ClientHttp1StreamHandler.java:291)
	at org.apache.hc.core5.http.impl.nio.ClientHttp1StreamDuplexer.terminate(ClientHttp1StreamDuplexer.java:181)
	at org.apache.hc.core5.http.impl.nio.AbstractHttp1StreamDuplexer.shutdownSession(AbstractHttp1StreamDuplexer.java:165)
	at org.apache.hc.core5.http.impl.nio.AbstractHttp1StreamDuplexer.onException(AbstractHttp1StreamDuplexer.java:407)
	at org.apache.hc.core5.http.impl.nio.AbstractHttp1IOEventHandler.exception(AbstractHttp1IOEventHandler.java:90)
	at org.apache.hc.core5.http.impl.nio.ClientHttp1IOEventHandler.exception(ClientHttp1IOEventHandler.java:41)
	at org.apache.hc.core5.reactor.InternalDataChannel.onException(InternalDataChannel.java:181)
	at org.apache.hc.core5.reactor.InternalChannel.handleIOEvent(InternalChannel.java:55)
	at org.apache.hc.core5.reactor.SingleCoreIOReactor.processEvents(SingleCoreIOReactor.java:176)
	at org.apache.hc.core5.reactor.SingleCoreIOReactor.doExecute(SingleCoreIOReactor.java:125)
	at org.apache.hc.core5.reactor.AbstractSingleCoreIOReactor.execute(AbstractSingleCoreIOReactor.java:92)
	at org.apache.hc.core5.reactor.IOReactorWorker.run(IOReactorWorker.java:44)
	... 1 common frames omitted
Caused by: java.lang.IndexOutOfBoundsException: Range [0, 0 + 4279) out of bounds for length 2048
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckFromIndexSize(Preconditions.java:118)
	at java.base/jdk.internal.util.Preconditions.checkFromIndexSize(Preconditions.java:397)
	at java.base/java.util.Objects.checkFromIndexSize(Objects.java:437)
	at java.base/java.nio.HeapByteBuffer.get(HeapByteBuffer.java:187)
	at io.camunda.zeebe.client.impl.http.TypedApiEntityConsumer$RawApiEntityConsumer.consumeData(TypedApiEntityConsumer.java:178)
	at io.camunda.zeebe.client.impl.http.ApiEntityConsumer.data(ApiEntityConsumer.java:91)
	at org.apache.hc.core5.http.nio.entity.AbstractBinDataConsumer.consume(AbstractBinDataConsumer.java:77)
	at org.apache.hc.core5.http.nio.support.AbstractAsyncResponseConsumer.consume(AbstractAsyncResponseConsumer.java:134)
	at org.apache.hc.client5.http.impl.async.HttpAsyncMainClientExec$1.consume(HttpAsyncMainClientExec.java:276)
	at org.apache.hc.core5.http.impl.nio.ClientHttp1StreamHandler.consumeData(ClientHttp1StreamHandler.java:261)
	at org.apache.hc.core5.http.impl.nio.ClientHttp1StreamDuplexer.consumeData(ClientHttp1StreamDuplexer.java:354)
	at org.apache.hc.core5.http.impl.nio.AbstractHttp1StreamDuplexer.onInput(AbstractHttp1StreamDuplexer.java:328)
	at org.apache.hc.core5.http.impl.nio.AbstractHttp1IOEventHandler.inputReady(AbstractHttp1IOEventHandler.java:64)
	at org.apache.hc.core5.http.impl.nio.ClientHttp1IOEventHandler.inputReady(ClientHttp1IOEventHandler.java:41)
	at org.apache.hc.core5.reactor.InternalDataChannel.onIOEvent(InternalDataChannel.java:143)
	at org.apache.hc.core5.reactor.InternalChannel.handleIOEvent(InternalChannel.java:51)
	... 5 common frames omitted

```

</p>
</details>

**Environment:**
- OS: MacOS
- Zeebe Version: 8.7.0-SNAPSHOT
- Configuration: all defaults

--- a/clients/java/src/main/java/io/camunda/zeebe/client/impl/http/ApiEntityConsumer.java
+++ b/clients/java/src/main/java/io/camunda/zeebe/client/impl/http/ApiEntityConsumer.java
@@ -72,7 +72,7 @@ final class ApiEntityConsumer<T> extends AbstractBinAsyncEntityConsumer<ApiEntit
       final boolean isResponse =
           String.class.equals(type)
               && SUPPORTED_TEXT_CONTENT_TYPES.stream().anyMatch(t -> t.isSameMimeType(contentType));
-      entityConsumer = new RawApiEntityConsumer<>(isResponse);
+      entityConsumer = new RawApiEntityConsumer<>(isResponse, maxCapacity);
     }
   }
 

--- a/clients/java/src/main/java/io/camunda/zeebe/client/impl/http/TypedApiEntityConsumer.java
+++ b/clients/java/src/main/java/io/camunda/zeebe/client/impl/http/TypedApiEntityConsumer.java
@@ -146,13 +146,15 @@ public interface TypedApiEntityConsumer<T> {
   class RawApiEntityConsumer<T> implements TypedApiEntityConsumer<T> {
 
     private final boolean isResponse;
+    private final int maxCapacity;
 
     private byte[] body = new byte[1024];
 
     private int bufferedBytes;
 
-    public RawApiEntityConsumer(final boolean isResponse) {
+    public RawApiEntityConsumer(final boolean isResponse, final int maxCapacity) {
       this.isResponse = isResponse;
+      this.maxCapacity = maxCapacity;
     }
 
     @Override
@@ -173,7 +175,11 @@ public interface TypedApiEntityConsumer<T> {
       final int offset = bufferedBytes;
       bufferedBytes += src.remaining();
       if (body.length < bufferedBytes) {
-        body = Arrays.copyOf(body, body.length + 1024);
+        if (bufferedBytes > maxCapacity) {
+          throw new IllegalArgumentException(
+              "The message size exceeds the maximum allowed size of " + maxCapacity);
+        }
+        body = Arrays.copyOf(body, bufferedBytes);
       }
       src.get(body, offset, src.remaining());
     }

